### ISSUE TITLE
Multiple endpoints

### ISSUE THREAD

**Issue by Symr81** [Geuthur/EVE-Alert#19]
Have you thought about setting up socket communication between multiple endpoints?
In this case, even Alpha account could participate in monitoring the local.
This way, there could be an endpoint that only alerts, while the others monitoring.

**Comment by Geuthur** [comment]
Did you mean if you have more then 1 Alert Program active?

I am not familiar with the Socket System, so I didn't think about it, but it would be nice if there was one. I think it would be useful because then you could use it in any way to give the link to someone else, and they would be able to see the status of a specific location.

**Comment by Symr81** [comment]
yes exactly

I'm not very good at it either, it's not my main p.language, but I'll try.
Similar to AlertAgent.alarm_detection, (AlertAgent.send_notify)

trying to make some demo on localhost first

but Christmas is coming... who knows how much time I'll have :)

PS: This one is good example: https://github.com/katmfoo/python-client-server

**Comment by Geuthur** [comment]
I have already done this in a streamdeck app.. so little exp is there

Yes christmas is coming but I will make a future Project List and add it to the list as Suggestion

If you will do it let me know if you want ;)

**Comment by Symr81** [comment]
![image](https://github.com/user-attachments/assets/ad0717e3-5f08-4188-83d2-702ed7040034)

I made a little demo and improvements
- it automatically decides whether to be in client or server mode
- you can test the server side from the client side
- the server can also send messages back to the client (currently just for testing)
- the online image has become a heartbeat like (flashing)

I think it's fine for me, I can monitor the local from an alpha account.

**Comment by Geuthur** [comment]
Sounds Awesome.

This will open a better or easier way to give Alert to all People that only want to use it as Client.
I will write later a Client Only Tool that only recieve the Socket Adress and play a alert when the Server has detected a Neutral.

If you want you can make a Pull Request and we can look together over it.
Maybe i can add the Single Client tool to the written Test and work on a release one.


P.S 
I have a Vision that maybe will be good what do you thing??

Example:

Alertermode: Same Use as Normal only different that you send the Signal to the Socket Server
Recievermode: Recieve only Signal and Play  Sound if the Socket has a True Signal
Socket Server: Recieve and Sendout the Status of the Signal

Player 1 use Alert as Alerter from 10:00 to 15:00 all clients will recieve alerts from Player 1
Player 1 is now offline and Player 2 activate the Alerter from his account to the Socket from 15:00 to 20:00
Player 3-4-5 only recieve the Signal and play alert if someone is in local

This is a way that you can use the Alert as a Server Side System and all People with the Socket Adress can Recieve the Alert.
To Activate the Alertermode you can use the Same adress and maybe with a Password you can Activate the Alertermode.

**Comment by Symr81** [comment]
![image](https://github.com/user-attachments/assets/30103751-00b7-4ff4-b336-b0f0ee5c0019)

That sounds good, I was just about to draw something similar.

**Comment by Geuthur** [comment]
Exactly what i am thinking about :D
Ok, I will make a new Branch where I will develop this.

**Comment by Symr81** [comment]
Btw, is it legal? If u do it right just a few steps from a completly online INTEL system, which one is running from web host… Every alliance can register, ( or buy it via plex :) ) and all of the members can report and watch a complete region… omg

**Comment by Geuthur** [comment]
If you make money with it I think not but it is also not legal to use a Tool like this ;) 

This means if you will be detected the Account will be Banned 

But to detect this it is very difficult cause it Does not Read the ram or something from eve it make Screenshot from the Desktop area and search for the icon

**Comment by Symr81** [comment]
Yeah i know, i made a similar console app in '24-March using opencv as well... alerting via sound and image and counting the neuts for serious alert! :D

![image](https://github.com/user-attachments/assets/899c0bb5-f252-443b-bfd7-140c6a220f6c)

**Comment by Geuthur** [comment]
My First Script working perfectly..

I will do some stuff over the Christmas but very slow :)
But i think i will get this work in a few weeks.

![Screenshot 2024-12-20 092559](https://github.com/user-attachments/assets/6af085e7-675c-4ec0-a7d5-f9cab40f1001)

The Admin Access will give him right to send to the server and give alerts ;)
the rest will only be a listener

**Comment by Geuthur** [comment]
https://github.com/Geuthur/EVE-Alert-Opensource/releases/tag/v0.9.0a1

@Symr81 
The Server Version can be used as old tool or as Server Tool.
The Client Version is Reciever only and will connect to Server and play sound if enemy is detected.

You need at least 1 Server Client that you can Recieve Alerts with Client.

If a Server is Online and you want to have a second Alerter you also use the Server Tool and Disable Server Mode on Settings and Start Socket Server then you only send Alerts to the Server.
At the moment only password test123 works.

Let me know what you think.

**Comment by Symr81** [comment]
Nice work! That was fast! :) I tested it a bit.

Some feedback:
- maybe we could give the server a name setting and then the client could receive from whom the signal is coming. e.g. Tama-Nourvukaiken gate, or just F-XWIN (maybe limit to 10 char)
- if the server stops, the client (or the Alerter) should notice this, maybe some heartbeat message will be needed in the background
- in Server Mode or Alerter mode, it would be useful we can turn off the sound alert.
- the client mode is started with auto connect, it can be an option for that, or dont do that in final release (for example the IP is bad)
- how do you imagine the client who is in receiver mode will only be able to connect to one server? Maybe it would be useful to not have to start multiple clients (but that doesn't matter at all)
- we could maybe make the server smarter with some REST-API function. The notifications could be uploaded to a site in some JSON format. That way, a client is not necessarily needed (Receiver mode) or even a python discod bot could be the output. (there are many example projects here on GIT as well)

**Comment by Geuthur** [comment]
Hey,

Point 1 I can add
Point 2 Is already implemented the problem is that I don't found a 100% Solution for it cause there is some issues but the client get a notice if no alert will send longer then 5 secs
Point 3 is also possible

The rest can be added in future but for the first version i think it is enough :P

![Screenshot 2025-01-02 100839](https://github.com/user-attachments/assets/536f07b3-2fbd-4be1-b98e-ee9446a618c6)

**Comment by Geuthur** [comment]
https://github.com/Geuthur/EVE-Alert-Opensource/releases/tag/v0.9.0a2

Here is the new version with all the new stuff

Alert Mute
Server Name

Not Implemented:;
Auto Connect
Multiple Connects
Rest API

**Comment by Symr81** [comment]
Sup! Nice!

- there could be another debug message besides the "something went wrong" message, it's hard to find out exactly what's wrong, I get errors quite often in alerter mode, btw after that the whole detection was crashed (possibly is vision.py) and I don't get any more enemy detections
- I actually thought it would be a good idea to introduce the server name property so that the "Alert" message could contain this name
- the AlertAgent class could have an enemy/faction counter so I can pass the number on to the discord bot, maybe `len(enemy) in the AlertAgent - vision_thread method`
- This way the server can compose a message for e.g. the discord bot. This is a real alternative to the listener client, as many people have dynamic IPs from their provider or play via a VPN.

- discord bot

```
import discord

# Global variables to be set externally and bot to be triggered
token = ''
channelId = ''
message = ''

intents = discord.Intents.default()

# initializing discord client with default intents
client = discord.Client(intents=intents)

# event triggered when the bot is ready
@client.event
async def on_ready():
    channel = client.get_channel(int(channelId)) # find the channel with the channel ID
    await channel.send(message) # send the message from the global variable
    await client.close() # close the client to stop the client blocking code

# method to be executed from an external file
def BotRun():
    client.run(token)
```

- one more thing, I think the server name and mute option are not server settings, but rather program, I would put them under the root section in the json, but this is just a suggestion

**Comment by Geuthur** [comment]
If there is something wrong message mostly it has more Information in the Logs

I have tested the system and all Works fine for me.

There are  many improvments Things that will take a lot of time ;)

The discord Implementation is a final Thing

**Comment by Geuthur** [comment]
- [x]  Client Message with ServerName Prefix
- [x]  Better Logging System
- [ ]  Alarm Counter?
- [ ]  Discord Implementation

Todo List

**Comment by Symr81** [comment]
That's all I found, but I'm sure the entire CV is stopped. (I test with a local image that I slide into the tested area.)
```
[2025-01-03 13:10:46] ERROR alertmanager _play_sound 396: Error playing audio: Error querying device -1
[2025-01-03 13:11:18] ERROR alertmanager _play_sound 396: Error playing audio: Error querying device -1
```

**Comment by Geuthur** [comment]
This Error I Think a Hardware Issue maybe cause there is no soundcard pluged in?

**Comment by Symr81** [comment]
Exactly, I just figured it out. The second PC is a Gigabyte Cube, and if you don't have headphones plugged in, there's no player, at least that's how Windows handles it, even if the sound could go out of the HDMI... 

another development opportunity.... :D
btw the mute option is also good for handling this, it just needs to be detected...

**Comment by Geuthur** [comment]
I have seen that when this error occurs the Script will Stop I will remove that.
I will also Change The Error Text.

This will solve both problems :D

**Comment by Geuthur** [comment]
https://github.com/Geuthur/EVE-Alert-Opensource/releases/tag/v0.9.0a3

I think now all must be work except the discord implementation if it is so and it works well in a period.
Then i will continue with a Discord Implementation.

**Comment by Symr81** [comment]
wow, that's a quick job
I'll test it right now.

In my own sandbox, I treated external signals separately from internal ones in the AlertAgent run routine, but I realized that this is not good either. If an alerter signals and hits the cooldown (locked()/release()), it can miss the signals of another alerter.
Therefore, the dictionary (alarm_counter) should be treated and counted based on the server name, not just "Enemy". Am I right?

**Comment by Geuthur** [comment]
I am not sure what do you mean with the Cooldown but if you have Enemy in different Systems you receive both and if the Sound is on Cooldown then you have already heard it and after 10 secs it resets so normally you have already the attention and you can read the message that more then Jita is Alerted.

I have also tested it and have seen that the name is not displaying on the Broadcaster Mode this will fix in the next version.
